File: ios/Sources/BubblSDK/Transport.swift
import Foundation

public struct BubblHTTPRequest: Sendable, Equatable {
    public let method: String
    public let url: URL
    public let headers: [String: String]
    public let body: Data?

    public init(method: String, url: URL, headers: [String: String] = [:], body: Data? = nil) {
        self.method = method
        self.url = url
        self.headers = headers
        self.body = body
    }
}

public struct BubblHTTPResponse: Sendable, Equatable {
    public let statusCode: Int
    public let data: Data
    public let headers: [String: String]

    public init(statusCode: Int, data: Data = Data(), headers: [String: String] = [:]) {
        self.statusCode = statusCode
        self.data = data
        self.headers = headers
    }
}

public protocol BubblHTTPTransport: Sendable {
    func send(_ request: BubblHTTPRequest) async throws -> BubblHTTPResponse
}

public struct URLSessionBubblHTTPTransport: BubblHTTPTransport {
    private let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    public func send(_ request: BubblHTTPRequest) async throws -> BubblHTTPResponse {
        var urlRequest = URLRequest(url: request.url)
        urlRequest.httpMethod = request.method
        urlRequest.httpBody = request.body
        urlRequest.timeoutInterval = 30

        for (name, value) in request.headers {
            urlRequest.setValue(value, forHTTPHeaderField: name)
        }

        let (data, response) = try await session.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw BubblError.invalidResponse("Response was not HTTP.")
        }

        let headers = httpResponse.allHeaderFields.reduce(into: [String: String]()) { result, item in
            guard let name = item.key as? String else { return }
            result[name] = String(describing: item.value)
        }

        return BubblHTTPResponse(statusCode: httpResponse.statusCode, data: data, headers: headers)
    }
}

enum BubblTransportMap {
    static let sdkVersion = "3.0.0"
    static let platform = "ios"

    static let runtimeAuthHeader = "x-api-key"
    static let ingestAuthHeader = "ApiKey"
    static let dashboardAuthHeader = "ApiKey"

    static let refreshGeofencePath = "/api/check-geofence"
    static let refreshPushPath = "/api/check-push"
    static let getConfigurationPath = "/api/get-config"

    static let registerDevicePath = "/api/device-registerd/create"
    static let bootBatchPath = "/api/device-data"
    static let trackEventPath = "/api/activities"
    static let updateSegmentsPath = "/api/segments"
    static let submitSurveyResponsePath = "/api/survey-response"
    static let trackGeofenceBatchPath = "/api/geofence-data"

    static func transmissionBaseURL(_ config: BubblConfig) -> URL {
        if let url = config.transmissionBaseUrl ?? config.runtimeBaseUrl {
            return url
        }

        switch config.environment {
        case .development, .nightly:
            return URL(string: "https://nightly.transmission.bubbl.tech")!
        case .staging:
            return URL(string: "https://staging.transmission.bubbl.tech")!
        case .production:
            return URL(string: "https://transmission.bubbl.tech")!
        }
    }

    static func runtimeBaseURL(_ config: BubblConfig) -> URL {
        transmissionBaseURL(config)
    }

    static func ingestBaseURL(_ config: BubblConfig) -> URL {
        if let url = config.ingestBaseUrl {
            return url
        }

        switch config.environment {
        case .development, .nightly:
            return URL(string: "https://nightly.ingest.bubbl.tech")!
        case .staging:
            return URL(string: "https://staging.ingest.bubbl.tech")!
        case .production:
            return URL(string: "https://ingest.bubbl.tech")!
        }
    }

    static func transmissionDistanceMiles(publicDistanceMeters: Int) -> Double {
        Double(max(publicDistanceMeters, 1)) / 1_609.344
    }
}

File: android/src/main/kotlin/tech/bubbl/sdk/Transport.kt
package tech.bubbl.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL

data class BubblHttpRequest(
    val method: String,
    val url: String,
    val headers: Map<String, String> = emptyMap(),
    val body: String? = null
)

data class BubblHttpResponse(
    val statusCode: Int,
    val body: String,
    val headers: Map<String, List<String>> = emptyMap()
)

fun interface BubblHttpTransport {
    suspend fun send(request: BubblHttpRequest): BubblHttpResponse
}

class UrlConnectionBubblHttpTransport : BubblHttpTransport {
    override suspend fun send(request: BubblHttpRequest): BubblHttpResponse = withContext(Dispatchers.IO) {
        val connection = URL(request.url).openConnection() as HttpURLConnection
        connection.requestMethod = request.method
        connection.connectTimeout = 30_000
        connection.readTimeout = 30_000
        connection.doInput = true

        request.headers.forEach { (name, value) ->
            connection.setRequestProperty(name, value)
        }

        request.body?.let { body ->
            connection.doOutput = true
            connection.outputStream.use { stream ->
                stream.write(body.toByteArray(Charsets.UTF_8))
            }
        }

        val status = connection.responseCode
        val responseBody = runCatching {
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
        }.getOrDefault("")

        BubblHttpResponse(
            statusCode = status,
            body = responseBody,
            headers = connection.headerFields
                .orEmpty()
                .filterKeys { it != null }
                .mapKeys { it.key.orEmpty() }
                .mapValues { it.value.orEmpty() }
        )
    }
}

internal object BubblTransportMap {
    const val sdkVersion = "3.1.3"
    const val platform = "android"

    const val runtimeAuthHeader = "x-api-key"
    const val ingestAuthHeader = "ApiKey"
    const val dashboardAuthHeader = "ApiKey"

    const val refreshGeofencePath = "/api/check-geofence"
    const val refreshPushPath = "/api/check-push"
    const val getConfigurationPath = "/api/get-config"

    const val registerDevicePath = "/api/device-registerd/create"
    const val bootBatchPath = "/api/device-data"
    const val trackEventPath = "/api/activities"
    const val updateSegmentsPath = "/api/segments"
    const val submitSurveyResponsePath = "/api/survey-response"
    const val trackGeofenceBatchPath = "/api/geofence-data"

    fun transmissionBaseUrl(config: BubblConfig): String =
        config.transmissionBaseUrl ?: config.runtimeBaseUrl ?: when (config.environment) {
            BubblEnvironment.Development,
            BubblEnvironment.Staging -> "https://staging.transmission.bubbl.tech"
            BubblEnvironment.Production -> "https://transmission.bubbl.tech"
        }

    fun runtimeBaseUrl(config: BubblConfig): String = transmissionBaseUrl(config)

    fun ingestBaseUrl(config: BubblConfig): String =
        config.ingestBaseUrl ?: when (config.environment) {
            BubblEnvironment.Development,
            BubblEnvironment.Staging -> "https://staging.ingest.bubbl.tech"
            BubblEnvironment.Production -> "https://ingest.bubbl.tech"
        }

    fun transmissionDistanceMiles(publicDistanceMeters: Int): Double =
        publicDistanceMeters.coerceAtLeast(1) / 1609.344
}
